src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c

```diff
@@ -0,0 +1,1041 @@
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].maxOutputChannels = 0;
+
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].defaultLowInputLatency = (double)l->latency;
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].defaultLowOutputLatency = 0;
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].defaultHighInputLatency = (double)l->configured_latency;
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].defaultHighOutputLatency = 0;
+
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].defaultSampleRate = l->sample_spec.rate;
+
+    l_ptrHostApi->deviceCount ++;
+
+error:
+    PA_DEBUG(("Portaudio %s: End of list\n",__FUNCTION__));
+
+    pa_threaded_mainloop_signal(l_ptrHostApi->mainloop, 0);
+}
+
+/* This routine is called whenever the stream state changes */
+void PulseaudioStreamStateCb(pa_stream *s, void *userdata)
+{
+    const pa_buffer_attr *a;
+    char cmt[PA_CHANNEL_MAP_SNPRINT_MAX], sst[PA_SAMPLE_SPEC_SNPRINT_MAX];
+    assert(s);
+
+    /* printf("Portaudio [Pulseaudio (PulseaudioStreamStateCb)]: Stream STATE CHANGED: "); */
+
+    switch (pa_stream_get_state(s))
+    {
+    case PA_STREAM_TERMINATED:
+        /* printf("TERMINATED!\n"); */
+        break;
+
+    case PA_STREAM_CREATING:
+        /* printf("CREATING!\n"); */
+        break;
+
+    case PA_STREAM_READY:
+        /* printf("READY!\n"); */
+
+        /*fprintf(stderr, "PulseaudioStreamStateCb: Stream successfully created.\n");
+
+        if (!(a = pa_stream_get_buffer_attr(s)))
+            fprintf(stderr, "pa_stream_get_buffer_attr() failed: %s\n", pa_strerror(pa_context_errno(pa_stream_get_context(s))));
+        else {
+                fprintf(stderr, "PulseaudioStreamStateCb: Buffer metrics: maxlength=%u, tlength=%u, prebuf=%u, minreq=%u\n", a->maxlength, a->tlength, a->prebuf, a->minreq);
+        }
+
+        fprintf(stderr, "PulseaudioStreamStateCb: Using sample spec '%s', channel map '%s'.\n",
+                pa_sample_spec_snprint(sst, sizeof(sst), pa_stream_get_sample_spec(s)),
+                pa_channel_map_snprint(cmt, sizeof(cmt), pa_stream_get_channel_map(s)));
+
+        fprintf(stderr, "PulseaudioStreamStateCb: Connected to device %s (%u, %ssuspended).\n",
+                pa_stream_get_device_name(s),
+                pa_stream_get_device_index(s),
+                pa_stream_is_suspended(s) ? "" : "not ");*/
+        break;
+
+    case PA_STREAM_FAILED:
+    default:
+        PA_DEBUG(("Portaudio %s: FAILED '%s'\n",__FUNCTION__, pa_strerror(pa_context_errno(pa_stream_get_context(s)))));
+
+        break;
+    }
+}
+
+void PulseaudioStreamUnderflowCb(pa_stream *s, void *userdata)
+{
+    PaPulseaudioStream *stream = (PaPulseaudioStream*)userdata;
+
+    stream->underflows++;
+
+    if (stream->underflows >= 6 && stream->latency < 2000000)
+    {
+        stream->latency = (stream->latency * 3) / 2;
```

Comment:
Actually only writing is driven by callback and reading blocking. Hmm.. how do I explain this. You are almost correct but you got it little bit wrong.
Reported latency by PulseAudio is latency from reading/writing ALSA which is reported by soundchip and kernel. PulseAudio undderrun is situation where we have too much latency (for keeping stream alive) for example we can't write fast enough to speakers.
So it's misleading because you can only estimate the latency from mixing to PulseAudio to ALSA to Chip to Speakers (Actually PulseAudio is as fast as Jack but that is another story). If there is no underrun we just keep making CB and dancers are :dancers:.
PulseAudio does callbacks even if we don't feed anything out or in then we have undderun. It tries to keep you up and running in selected sample rate and sample mode (S16, Float). What we feed through the Portaudio API is just as fast as we get it through PulseAudio CB and then to the Mixxx or Pulseaudio asks for samples through Portaudio (As ALSA does) and then it feeds them to ALSA.
You should refer this one: https://github.com/illuusio/linux-audio-example where I tried to make simple examples how different Linux Audio APIs work. It's work in process but contain Pulseaudio and Portaudio so you can see how they should work.